posts/2024-01-31-Performance-Doris_Amoakohene/index.qmd

```diff
@@ -0,0 +1,240 @@
+
+*Lines:*
+
+3. “Before”: Indicates performance before fixing the regression.
+
+4. “Fixed”: Shows improved performance after fixing.
+
+5. “Regression”: Represents an ideal or target performance level.
+
+In the graph, as data size (N) increases, there’s an initial increase in median time, but after addressing the regression issue, there is a significant reduction in the median time, indicating improved performance(fix). 
+The regression issue was successfully addressed.
+
+
+# Example Two
+
+In the specific case of issue *#5366*, there was a significant slowdown in the performance of data.table's time-based rolling aggregation compared to pandas' rolling aggregation. The cause of this regression was identified to be related to the addition of the `snprintf` function in the `assign.c` file. To address this issue, a fix was implemented by creating the `targetDesc` function and adding the `snprintf` function in the `assign.c` file. This fix resolved the regression and improved the performance of the time-based rolling aggregation in data.table. 
+
+```{r,warning = FALSE, message = FALSE}
+atime.list.5366 <- atime::atime_versions(
+pkg.path=tdir,
+pkg.edit.fun = pkg.edit.fun,
+  N= 10^seq(1,4),
+  setup={ 
+    set.seed(123L)
+    dt <- data.table(
+      id = seq_len(N),
+      val = rnorm(N))
+    dt
+  },
+ expr=data.table:::`[.data.table`(dt, , .(vs = (sum(val))), by = .(id)),
+  "Before"="be2f72e6f5c90622fe72e1c315ca05769a9dc854",
+  "Regression"="e793f53466d99f86e70fc2611b708ae8c601a451", 
+  "Fixed"="58409197426ced4714af842650b0cc3b9e2cb842") 
+```
+
+
+## Significantly slower performance time-based rolling
+
+![Plot showing the 3 branches(Regression, Fixed and Before) of Issues5366](https://github.com/DorisAmoakohene/Efficiency-and-Performance-Test.RData.table/blob/main/atime.list.5366.png)
+Graph overview:
+The graph depicts the performance improvement over time after fixing a regression issue
+Y-axis: Represents time in milliseconds.
+X-axis: Represents the sample size N (ranging from 10 to 10,000).
+
+*Three lines:*
+
+“Before Fixed”: Indicates performance before fixing the regression.
+“Fixed”: Represents improved performance after fixing.
+“Regression”: Shows the regression issue.
+The Regression line initially increases steeply with larger N, indicating slow performance.
+After fixing, line shows significantly reduced processing time.
+
+In summary, the graph visually demonstrates how fixing the regression issue led to improved performance in time-based rolling operations. The Fixed line represents the desired outcome, showing faster processing times for larger sample sizes. 
+
+
+# **Conclusion**
+
+In this blog post, we have delved into the use of the atime code to compare the asymptotic time and memory usage of different versions of the data.table package. Specifically, we explored the comparisons between the "Before," "Regression," and "Fixed" versions, as well as different versions implementing the same computation.
```

Comment:
please add a description of the github action, and how data.table is now using this to review PRs like this https://github.com/Rdatatable/data.table/pull/5427#issuecomment-2075471806

Reply:
addressed